File: LeetCode/198. House Robber.cpp
// Author : @AdarshGaur

class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n == 1) return nums.back();
        int l1 = nums[1], l2 = nums[0], l3 = 0;
        for(int i=2; i<n; i++){
            int t = l1;
            l1 = max(l2, l3) + nums[i];
            l3 = l2;
            l2 = t;
        }
        return max(l1, l2);
    }
};

File: LeetCode/198. House Robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        for(int i=0; i<n; i++){
            int k = i - 2;
            int m = i - 3;
            if(k >= 0 && m >= 0){
                dp[i] = Math.max(dp[k], dp[m]) + nums[i];
            }else if(k >= 0){
                dp[i] = dp[k] + nums[i];
            }else if(m >= 0){
                dp[i] = dp[m] + nums[i];
            }else{
                dp[i] = nums[i];
            }
        }
        return n > 1 ? Math.max(dp[n-2], dp[n-1]) : dp[n-1];
    }
}
